Wikipedia/View Controllers/WebView/NativeSectionTitleOverlays/WMFSectionTitlesViewController.m

```diff
@@ -0,0 +1,231 @@
+    NSUInteger newTopLabelIndex = -1;
+
+    for (NSUInteger thisIndex = 0; thisIndex < self.models.count; thisIndex++) {
+        WMFTitleOverlayModel* m = self.models[thisIndex];
+        CGFloat thisOffset      = m.yOffset;
+        if (offsetY > lastOffset && offsetY <= thisOffset) {
+            newTopLabelIndex = thisIndex - 1;
+            break;
+        } else if ((thisIndex == (self.models.count - 1)) && (offsetY > thisOffset)) {
+            newTopLabelIndex = thisIndex;
+            break;
+        }
+        lastOffset = thisOffset;
+    }
+
+    if (newTopLabelIndex != -1) {
+        if (newTopLabelIndex != self.indexOfNativeTitleLabelNearestTop) {
+            self.indexOfNativeTitleLabelNearestTop = newTopLabelIndex;
+            [self updateTopStaticTitleLabelText];
+        }
+    }
+}
+
+- (void)nudgeTopStaticTitleLabelIfNecessaryForScrollOffsetY:(CGFloat)offsetY {
+    NSUInteger pusherIndex  = self.indexOfNativeTitleLabelNearestTop + 1;
+    CGFloat distanceToPushY = 0;
+    if (pusherIndex < (self.models.count)) {
+        WMFTitleOverlayModel* pusherTitleLabel = self.models[pusherIndex];
+        if (pusherTitleLabel.sectionId > 0) {
+            CGFloat topmostHeaderOffsetY  = pusherTitleLabel.yOffset;
+            CGRect staticLabelPseudoRect  = CGRectMake(0, 0, 1, self.topStaticNativeTitleLabel.frame.size.height);
+            CGRect topmostLabelPseudoRect = CGRectMake(0, topmostHeaderOffsetY - offsetY, 1, 1);
+            if (CGRectIntersectsRect(staticLabelPseudoRect, topmostLabelPseudoRect)) {
+                distanceToPushY = staticLabelPseudoRect.size.height - topmostLabelPseudoRect.origin.y;
+            }
+        }
+    }
+    self.topStaticNativeTitleLabelTopConstraint.constant = -distanceToPushY;
+}
+
+- (void)updateTopStaticTitleLabelText {
+    self.topStaticNativeTitleLabel.alpha = 1.0;
+    WMFTitleOverlayModel* m = self.models[self.indexOfNativeTitleLabelNearestTop];
+    self.topStaticNativeTitleLabel.text      = m.title;
+    self.topStaticNativeTitleLabel.sectionId = m.sectionId;
+}
+
+#pragma mark Section title and title location determination
+
+- (id)getJSONFromWebViewUsingFunction:(NSString*)jsFunctionString {
+    NSString* jsonString = [self.webView stringByEvaluatingJavaScriptFromString:jsFunctionString];
+    NSData* jsonData     = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
+    NSError* error       = nil;
+    id result            = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
+    return (error) ? nil : result;
+}
+
+static NSString* const WMFJSGetSectionTitlesJSON =
+    @"(function(){"
+    @"  var nodeList = document.querySelectorAll('h1.section_heading');"
+    @"  var nodeArray = Array.prototype.slice.call(nodeList);"
+    @"  nodeArray = nodeArray.map(function(n){"
+    @"    var rect = n.getBoundingClientRect();"
+    @"    return {"
+    @"        anchor:n.id,"
+    @"        sectionId:n.getAttribute('sectionId'),"
```

Comment:
just because i don't want to take this for granted, we should verify none of these values are unexpected (e.g. `undefined` or `null`) and return an explicit error (which we check for in native land). in cases where we rely on the DOM, we should expect the unexpected